### PR TITLE
tsdb/agent: Fix deadlock from simultaneous GC and write

### LINKED ISSUE #10151: Prometheus Agent Stops Scraping all its Targets after sometime without any error logs.
Prometheus with --enable-feature=agent Stops Scraping all its Targets after sometime and when we check in terminal logs there is no such logs related scraping stop. 
Is anyone has any idea about this ???
Image : prometheus:v2.32.0
Stop Targets 
image

### PR DESCRIPTION
**Pull request by rfratto**
This commit fixes a potential deadlock where storing in-memory series references could deadlock with a WAL GC cycle.

I've tried by best to write a test to catch this. While I did manage to reproduce it locally with the test, it might cause false positives (i.e., test passes when it shouldn't) if your machine is fast enough. 

Fixes #10151 

Signed-off-by: Robert Fratto <robertfratto@gmail.com>

### CODE REVIEW

**Comment by roidelapluie** [comment]
Would the following pattern work for improving the catch rate?

https://github.com/prometheus/prometheus/blob/49f06955f446d4857e7c0fc65ce80dbd74879039/discovery/manager_test.go#L1422

https://github.com/prometheus/prometheus/blob/49f06955f446d4857e7c0fc65ce80dbd74879039/discovery/manager_test.go#L1424

https://github.com/prometheus/prometheus/blob/49f06955f446d4857e7c0fc65ce80dbd74879039/discovery/manager_test.go#L1449

Basically we create go routines but they execute only when we close the start channel. That increase the number of things we do in parallel.

**Comment by roidelapluie** [comment]
cc @juliusv  @beorn7 this might be a candidate for 2.32.x, 2.33.0 or a 2.33.1 release.

**Comment by rfratto** [comment]
> Would the following pattern work for improving the catch rate?

Thanks! I'll give it a shot when I'm back from PTO tomorrow. I was able to catch the error completely consistently on my machine, but I'm not confident it'll work for others as consistently and using an existing pattern would be preferable.

**Comment by rfratto** [comment]
> Basically we create go routines but they execute only when we close the start channel. That increase the number of things we do in parallel.

Thanks, this was useful and also showed that concurrent GCs can also deadlock (even though GCs can't currently happen concurrently, it's probably better to be defensive here). 

I think this is probably ready for a look cc @codesome.

**Comment by beorn7** [comment]
Not sure if this is urgent enough to go inte 2.32, with the 2.33 release being imminent.

@rfratto to facilitate merging this into 2.33, could you change this PR to be on top of the `release-2.33` branch?

**Comment by rfratto** [comment]
> @rfratto to facilitate merging this into 2.33, could you change this PR to be on top of the release-2.33 branch?

@beorn7 Done :+1: Do I need to do anything to get it back into main after 2.33?

**Comment by beorn7** [comment]
> Do I need to do anything to get it back into main after 2.33?

I will do that for you once this is merged.

**Review by codesome — APPROVED** [review]
